### ISSUE TITLE
508-Defect-1 [SCREENREADER] When scheduling an appointment fails, the user is given no feedback and the error is not announced

### ISSUE THREAD

**Issue by noahgelman** [department-of-veterans-affairs/va.gov-team#26391]
[508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-0)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** Noah

## User Story or Problem Statement

Failing to schedule an appointment does not announce the error with a screen reader

### When submitting the form, it does not announce to the user the state

It does not announce to screen readers that the form is submitting. A loading spinner needs to be added that announces that the form is being submitted

### When attempting to schedule an appointment and it fails, the user is given no feedback and the error is not announced

Add the appropriate `role="alert"` to the error box

![couldnt](https://user-images.githubusercontent.com/984252/122903318-23ddcb00-d304-11eb-982c-1e73ecf585cb.PNG)

## Acceptance Criteria

- [ ] Add loading spinner to the end of the form to announce the form is submitting
- [ ] Add the appropriate `role="alert"` to the error box

## WCAG or Vendor Guidance (optional)

[https://www.w3.org/TR/WCAG21/#error-prevention-all](https://www.w3.org/TR/WCAG21/#error-prevention-all1

**Comment by jbalboni** [comment]
I merged the change to add an alert role and focus management for the error message. 

I have not been able to replicate the loading text not being announced, and we're following the same pattern that other forms and applications use on va.gov for that. I requested some help from @noahgelman on Slack to look into that.

**Comment by noahgelman** [comment]
@jbalboni The error message announcement is good. I'll check with Trevor about the button loading announcement

**Comment by laurenernest** [comment]
Also tracking in Slack support request https://dsva.slack.com/archives/CBU0KDSB1/p1625076737454700

**Comment by 1Copenut** [comment]
@laurenernest I retested this in VoiceOver, NVDA, and JAWS briefly. JAWS isn't announcing the alert as expected. I'm attaching an image below that shows the `div[role="alert"]` will need to be in the DOM on first page render and then we drop the Alert box into it if the appointment can't be scheduled. We'll retest promptly after this update.

---
![Screen Shot 2021-07-15 at 12 05 00 PM](https://user-images.githubusercontent.com/934879/125828730-79783cd8-4f9b-4a8a-bb96-7777c10116fc.png)

**Comment by jbalboni** [comment]
@1Copenut I was following the pattern used by other va.gov forms. Is using an alert here even the correct way to handle this? Could we just focus on the headline?

This pattern of adding a div with role="alert" appears to also be how validation errors work.

Also, to be clear, this is separate from the other issue that we were trying to get feedback on in here, which is the loading text being read out.

**Comment by 1Copenut** [comment]
@jbalboni 
> this is separate from the other issue that we were trying to get feedback on in here, which is the loading text being read out.

Thank you for the nudge on this other question. After talking with Noah, we believe the [Loading Button component](https://github.com/department-of-veterans-affairs/vets-website/blob/21260300681f449c564f53596d0a8302dd684e7b/src/platform/site-wide/loading-button/LoadingButton.jsx) is not reading out the `loadingText` consistently because of where the `role="status"` attribute is placed. If we move it to the button tag, it should smooth out the experience and announce for all screen readers:

```diff
! LoadingButton.jsx#11

const contents = isLoading ? (
  <>
    <i
      className="fa fa-spinner fa-spin"
      aria-hidden="true"
      role="presentation"
    />
    {loadingText && (
-     <span className="sr-only" role="status">
+     <span className="sr-only">
        {loadingText}
      </span>
    )}
  </> 
) : (
  children
);

return (
  <button
    className="usa-button"
    {...props}
    disabled={isLoading || disabled}
    onClick={onClick}
+   role="status"
  >
    {contents}
  </button>
);
```

**Comment by jbalboni** [comment]
@1Copenut Made that change here, though it requires ignoring one of the a11y linting rules: https://github.com/department-of-veterans-affairs/vets-website/pull/17888

**Comment by jbalboni** [comment]
@1Copenut @noahgelman The loading button change is in staging, if that still needs to be retested.

**Comment by 1Copenut** [comment]
@laurenernest & @jbalboni I was able to confirm the button's changed text is announcing correctly in NVDA and JAWS on staging. I couldn't make the error state throw when scheduling some appointments, but previous testing on localhost has me confident we've bracketed the issue.

@noahgelman Please add any concerns or thoughts here.

**Comment by laurenernest** [comment]
@jbalboni Let's move forward with deploying this change. I'll be sending for full 508 audit soon so we'll catch any other feedback there.

**Comment by jbalboni** [comment]
@laurenernest It should already be live, so I think we're all set.

**Comment by laurenernest** [comment]
Great! Closing ticket.
